### PR TITLE
[SPARK-27630][CORE] Properly handle task end events from completed stages

### PR DESCRIPTION
**Pull request by cxzl25**
## What changes were proposed in this pull request?
Track tasks separately for each stage attempt (instead of tracking by stage), and do NOT reset the numRunningTasks to 0 on StageCompleted.

In the case of stage retry, the `taskEnd` event from the zombie stage sometimes makes the number of `totalRunningTasks` negative, which will causes the job to get stuck.
Similar problem also exists with `stageIdToTaskIndices` & `stageIdToSpeculativeTaskIndices`.
If it is a failed `taskEnd` event of the zombie stage, this will cause `stageIdToTaskIndices` or `stageIdToSpeculativeTaskIndices` to remove the task index of the active stage, and the number of `totalPendingTasks` will increase unexpectedly.
## How was this patch tested?
unit test properly handle task end events from completed stages

### CODE REVIEW

**Comment by cxzl25** [comment]
In the production environment with the patch of [SPARK-11334], we encountered a negative number of ExecutorAllocationManager totalRunningTasks, causing the job to get stuck.
@sitalkedia @vanzin @jerryshao 
Could you please give some comments when you have time?
Thanks so much.

image

**Comment by dongjoon-hyun** [comment]
ok to test

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: IIRC two stage attempts can actually be active at the same time, so this would not be correct.

Maybe @squito remembers this better than I do.

**Review by squito — CHANGES_REQUESTED** [review]
I left some comments about the meaning of active stages and some suggestions for improving the clarity of the test.  BUT, then I started to wonder what the value of `ExecutorAllocationManager#totalRunningTasks` should actually be in these scenarios.  This change makes it ignore running-but-zombie stage attempts.  I don't actually think that is right.

I left my suggestions for the test anyway (whatever behavior we want, I think that test setup is good, we just need to change the asserts).

If we *do* want the behavior here, than the doc on `ExecutorAllocationManager#totalRunningTasks` should be updated to reflect that it now only includes the non-zombie tasksets.

Also I think this should get its own jira, as the prior jira was already used for a released change -- reusing the jira makes it hard to track what was fixed when.

**Comment by cxzl25** [comment]
Thank you for your suggestions and help. @squito 

```ExecutorAllocationListener``` stores the stage related information, the hash map key is the stage id, and there is no stage attempt id, the statistical information is for the current active stage.

https://github.com/apache/spark/blob/6c2d351f5466d42c4d227f5627bd3709c266b5ce/core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala#L648-L663

In SPARK-11334, I saw a description added to stageIdToNumRunningTask
>Number of running tasks per stage including speculative tasks.
>Should be 0 when no stages are active.

If the stage's attempts are all zombie, their information should not be counted to the current active stage.
I tried adding a little comment on the totalRunningTasks, but it might be simpler.

**Comment by squito** [comment]
> ExecutorAllocationListener stores the stage related information, the hash map key is the stage id, and there is no stage attempt id, the statistical information is for the current active stage.

well ... that doesn't mean that's the right behavior.  From the ExecutionAllocationManager's point of view, running tasks in a zombie taskset are still using up slots on executors, right?  Eg. suppose you have five zombie tasks, and then another taskset with two more tasks comes along -- shouldn't the executor allocation manager try to request more executors for two more task slots?  

> In SPARK-11334, I saw a description added to stageIdToNumRunningTask
>>Number of running tasks per stage including speculative tasks.
>>Should be 0 when no stages are active.
>If the stage's attempts are all zombie, their information should not be counted to the current active stage.

I wouldn't assume that "active" is used consistently in the codebase with the way I defined it above.  Unfortunately I think the only term used consistently is zombie vs. non-zombie. Indeed the part you quote is discussing stages, not stage attempts, so I really doubt it means the same thing.

**Comment by cxzl25** [comment]
@squito 
Make sense!
In this scenario, the executor allocation manager should try to request more executors for two more task slots.
How to modify this logic will be more appropriate? 
Do you have any better ideas

**Comment by Ngone51** [comment]
I agree with @squito 's opinion that running tasks in zombie TaskSet should also be counted for `ExecutorAllocationManager#totalRunningTasks`. Basing on this, I'm wondering that would it be ok if we just changing `stageIdToNumRunningTask` to `stageAttemptIdToNumRunningTask` ?  As `ExecutorAllocationManager` does not really care about the stage is active or not, but care about the running tasks across the stage attempts. @squito @cxzl25

**Comment by squito** [comment]
the problem is, the `StageCompleted` for a stage retry comes before all those tasks finish ... but meanwhile when that happens for a task failing 4 times, we intentionally want to set the stage total to 0 for SPARK-11334.  Those two seem to conflict each other.  Though to be honest, I don't understand why that fix was used in SPARK-11334 -- it seems like you should just wait for the task end events, and not just any `numRunning` counts just on `StageCompleted` events.

I'm amazed this hasn't been reported before, it must be causing dynamic allocation to perform poorly on any stage retry.  I'd like to make sure we understand this properly and look a bit more at the history here.

@tgravescs @abellina  I think you'll be interested in this one too.

**Comment by squito** [comment]
I dug into the history of this a bit -- its a bit scattered across multiple PRs for https://issues.apache.org/jira/browse/SPARK-11334, and work done under https://issues.apache.org/jira/browse/SPARK-11701 and elsewhere.

It seems to me like we want to undo one key part of the SPARK-11334 -- where we remove an entry from `stageIdToNumRunningTasks` when we get a stageCompleted:

https://github.com/apache/spark/blob/bcd3b61c4be98565352491a108e6394670a0f413/core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala#L698

part of that change was to defend against a missing TaskEnd event.  But I think we just have to make sure those events are delivered, I don't see a way for this to work otherwise*. You can't clear the count-per-stage(attempt) on stageEnd, because the stage can still have tasks running after a stageCompleted event.  I'm not sure if it matters if you track running tasks per stage or stage attempt, I think it might not.  One thing we'll have to figure out is when you clear entries from `stageIdToNumRunningTask` to avoid it growing indefinitely -- I guess you can remove if the numRunningTasks is 0 AND its been cleared from the other structures.

@cxzl25 would you like to try to make this change?  This is a complex part which will require some careful review.

* there is the problem that the EventBus drops events when its full, but hopefully that's been improved significantly recently.  And anyway, it could drop a StageCompleted as easily as it could drop a TaskEnd.

**Comment by tgravescs** [comment]
So the issue why its negative is the task end event comes after the new stage attempt has started which causes it to decrement the stageIdToNumRunningTask(stage) Map.  Yeah I'm surprised we didn't see that more but it would be timing dependent.
What @squito says makes sense to me. 
I was just looking a bit and I wonder if we have an issues with the stageIdToSpeculativeTaskIndices as well.  If the stageId gets put back in there, you could have an issue though it seems unlikely.  I wonder if it makes senes to look at using the stage attempt id for a few of these.   I think you could have similar problem with stageIdToTaskIndices if the stage attempt had started other tasks before you got the task end.  I would have to take a more thorough look to verify.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: We shall still ensure it's not < 0 ?

Comment (cxzl25) [reply core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: In theory, there is no negative number of tasks.
Just to prevent negative numbers and to clear them in time.

Comment (jiangxb1987) [reply core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: In that case if we don't expect negative number of tasks to happen I would insert a assertion here, otherwise we can fail fast and print sufficient information.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: Which problem does this change aim to address? Do we have different expectation for 'regular' task and speculative task?

Comment (cxzl25) [reply core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: To simplify the calculation of totalRunningTasks (```add remove size```), instead of using the previous commit, use ```taskIdToNumLiveTasks = new mutable.HashMap[Long, Int] (add i++ i-- remove), values.sum```.
And in the case of many tasks can improve the sum calculation speed.
The final purpose is to prevent the totalRunningTasks from becoming negative, resulting in the job stuck, and can further improve the efficiency of applying for resources.

Comment (cxzl25) [reply core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: The taskid is actually not repeated.
Only the same taskid appears in some test cases, and the task has different types.
```scala
createTaskInfo(0, 0, "executor-1")
createTaskInfo(0, 0, "executor-2", true)
```

https://github.com/apache/spark/blob/cd4a284030e5142bdb405ff5b71735ac8cee2dde/core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala#L260-L267

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: If you only want to count the total number of active tasks why not count from `executorIdToTaskIds` ? This way you don't need to store any other data.

**Comment by jiangxb1987** [comment]
retest this please

**Comment by Ngone51** [comment]
I think similar problem also exists with `stageIdToTaskIndices` & `stageIdToSpeculativeTaskIndices`. Because, currently, we do not clear the `indices` when stage re-submit event comes(`SparkListenerStageSubmitted`). That's means `indices` would be used across multiple stage attempts until the stage complete. As a result of that, when task start/end events comes after stage re-submmited, we'll get wrong `indices`, which affect `pendingTasks()`/ `maxNumExecutorsNeeded()` counting finally. 

Maybe, we should really use `stageAttemptId` instead of `stageId` ?

**Comment by squito** [comment]
This is the mima error you're getting:

```
[error]  * the type hierarchy of object org.apache.spark.scheduler.SparkListenerSpeculativeTaskSubmitted is different in current version. Missing types {scala.runtime.AbstractFunction1}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("org.apache.spark.scheduler.SparkListenerSpeculativeTaskSubmitted$")
[error]  * method apply(Int)org.apache.spark.scheduler.SparkListenerSpeculativeTaskSubmitted in object org.apache.spark.scheduler.SparkListenerSpeculativeTaskSubmitted does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.scheduler.SparkListenerSpeculativeTaskSubmitted.apply")
[error]  * method copy(Int)org.apache.spark.scheduler.SparkListenerSpeculativeTaskSubmitted in class org.apache.spark.scheduler.SparkListenerSpeculativeTaskSubmitted does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.scheduler.SparkListenerSpeculativeTaskSubmitted.copy")
[error]  * method this(Int)Unit in class org.apache.spark.scheduler.SparkListenerSpeculativeTaskSubmitted does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.scheduler.SparkListenerSpeculativeTaskSubmitted.this")
```

its a developer api, so its OK to change it if need be, just update the rules in https://github.com/apache/spark/blob/master/project/MimaExcludes.scala#L38

(though I'm still trying to figure out whether we need to make that change or not ...)

File: core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala
Comment (squito) [inline core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala]: the name of this test is needs to be updated.  Maybe "properly handle task end events from completed stages"

File: core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala
Comment (squito) [inline core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala]: the last part of this comment needs to be updated -- I guess it should now be "we count all running tasks, from the zombie & non-zombie attempts"

File: core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala
Comment (squito) [inline core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala]: again, update comment "The tasks in the zombie attempt haven't completed yet, so we still count them"

File: core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala
Comment (squito) [inline core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala]: "We count completions from the first zombie attempt"

**Review by squito — COMMENTED** [review]
other than the mima update and changes to comments, I think this looks OK.  Would appreciate another careful review, though.

**Comment by vanzin** [comment]
Looks OK but please explain the fix, not the problem, in the PR title / description.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: IIRC, we receive only one `SparkListenerStageCompleted` event for multiple stage attempts. And, it's possible that `SparkListenerTaskEnd` comes from previous stage attempt after we've received `SparkListenerStageCompleted`. In that case, we need to clean up related structures for those stage attempts in `onTaskEnd`(if this is a final  finished task within corresponding stage attempt) as we already do in `onStageCompleted`.

Comment (cxzl25) [reply core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: In the stage retry process, `StageCompleted` will be sent first and then `StageSubmitted` will be sent, so `StageCompleted` will be received multiple times.
https://github.com/apache/spark/pull/24497#discussion_r280147031

Based on this statement, the corresponding stage attempt is cleaned up when `StageCompleted` is processed, so the `TaskEnd` after `StageCompleted` should have no relevant structure.
Unless `EventBus` is full, the `StageCompleted` is discarded.

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: Oh yes, I think you're right.

**Comment by squito** [comment]
lgtm

will leave a bit to let more people review

**Comment by tgravescs** [comment]
Just skimming through, would it make sense to make a local case class that is StageAttempt to replace  (stageId, stageAttemptId) and rename some of the maps to be stageAttempt.... rather than stageId? More for readability than anything.

**Comment by jiangxb1987** [comment]
> Just skimming through, would it make sense to make a local case class that is StageAttempt to replace (stageId, stageAttemptId) and rename some of the maps to be stageAttempt.... rather than stageId? More for readability than anything.

+1 Maybe we can extend the `ContextBarrierId`, it was not designed for general purpose but it seems to also fit in here (maybe just rename the class to a proper name)

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: Also update the variable names to `stageAttemptTo...` ?

**Comment by cxzl25** [comment]
retest this please

**Comment by tgravescs** [comment]
test this please

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: nit: stage Id -> stageAttempt

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: nit: stage -> stageAttempt

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: ditto.

**Review by Ngone51 — COMMENTED** [review]
LGTM, other than some nits.

**Comment by tgravescs** [comment]
case class and name changes look good other then nits mentioned, thanks.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: I'm still in doubt whether we shall keep track of all the running taskIds here, especially when the only use case can be covered by `stageIdToNumRunningTask.values.sum`

Comment (cxzl25) [reply core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: The previous commit version used ```executorIdToTaskIds.values.map(_.size).sum```, but in [SPARK-20286](https://issues.apache.org/jira/browse/SPARK-20286), ```executorIdToTaskIds``` was removed.

It seems that recording taskid is not important.
Now I changed ```stageIdToNumRunningTask``` to ```stageAttemptToNumRunningTask```.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala
Comment (squito) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: I think its worth adding a comment here along the lines of "do NOT remove stageAttempt from stageAttemptToNumRunningTasks, because the attempt may still have running tasks, even after another attempt for the stage is submitted."

**Review by squito — APPROVED** [review]
sorry for the long delay from me.  lgtm, I left one tiny comment.  can you also please update the PR descrption as vanzin requested earlier, to describe your solution, eg "Track tasks separately for each stage attempt (instead of tracking by stage), and do NOT reset the numRunningTasks to 0 on StageCompleted".  You can leave your problem description after that.

again since this is a tricky bit of code I'll wait till after the weekend to merge to give others another chance to look.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: I think we should not take tasks from previous stage attempts, which are already zombie,  into account, right ?

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: ditto.

File: project/MimaExcludes.scala
Comment (Ngone51) [inline project/MimaExcludes.scala]: Please do not forget to update this comment once you update the PR title.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala
Comment (cxzl25) [reply core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: In ```onStageCompleted``` processing, ```stageAttemptToNumTasks```, ```stageAttemptToTaskIndices``` removes the zombie stage, so pendingTasks does not calculate the zombie stage's task.

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: I see.

**Comment by Ngone51** [comment]
LGTM, again.

File: project/MimaExcludes.scala
Comment (cxzl25) [reply project/MimaExcludes.scala]: I have compared PR title and comment to be consistent.
What comment are missing?

Comment (Ngone51) [reply project/MimaExcludes.scala]: Oops, well enough.

**Comment by squito** [comment]
merged to master.  Thanks @cxzl25 !

**Pull request closed without merging**
